### PR TITLE
Implement OC, Coll, Resource from temp generation.

### PR DESCRIPTION
**Pull request by zancas**
@pjbreaux 
Issues:
Fixes: https://github.com/F5Networks/f5-common-python/issues/218

Problem: There's a lot of boilerplate involved in creating new concrete
subclasses.  This engine provides a tool to automatically generate that
boilerplate.

Analysis: The engine requires templates and JSON-config data as inputs
it produces Python boilerplate as a string as output.

Tests: These will be added in other stories.

Where to start:

Pass "RESTWalker.py" to a python.   You should get a print out of a Resource based on the json in the "`device_configs`" directory and the "`Resource.tmpl`" in the "template_library" directory.

### CODE REVIEW

File: utils/template_library/Resource.tmpl [hunk 1]
```diff
@@ -0,0 +1,8 @@
+
+
+class {{ klass }}(Resource):
+    def __init__(self, {{ container }}):
+        super({{ klass }}, self).__init__({{ container }})
+        self._meta_data['template_generated'] = True
+        self._meta_data['required_json_kind'] = {{ kind }}
+        self._meta_data['attribute_registry'] = {{ attr_reg_dict }}
```
Comment (pjbreaux) [inline utils/template_library/Resource.tmpl]: This is getting generated as follows:
```
class Pool(Resource):
    def __init__(self, Pools):
        super(Pool, self).__init__(Pools)
        self._meta_data['template_generated'] = True
        self._meta_data['required_json_kind'] = tm:ltm:pool:poolstate
        self._meta_data['attribute_registry'] =
```

The attribute_registry should be a dict.

File: utils/RESTWalker.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    def _format_org_collection(self, org_match, kind, raw_conf):
+        klass = org_match.group('OrgColl').capitalize()
+        OC_template = self.templates['OrganizingCollection']
+        import_dicts = self._build_import_dicts(raw_conf, klass)
+        config_dict = {'klass': klass,
+                       'kind': kind,
+                       'import_dicts': import_dicts}
+        OrgCollstr = OC_template.render(**config_dict)
+        imports_as_str = self.import_template.render(import_dicts=import_dicts)
+        imps_as_list = imports_as_str.split('\n')
```
Comment (pjbreaux) [inline utils/RESTWalker.py]: Try splitlines instead.

File: utils/RESTWalker.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        imps_as_list.append(
+            'from f5.bigip.resource import OrganizingCollection')
+        imps_as_list.sort()
+        imports = '\n'.join(imps_as_list)
```
Comment (pjbreaux) [inline utils/RESTWalker.py]: Is there a joinlines? Probably not.

File: utils/RESTWalker.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+        else:
+            raise Kindless(raw_conf)
+        return python_as_string
+
+    def _klassify(self, raw_klass):
+        caps = '_'.join([x.capitalize() for x in raw_klass.split('-')])
+        dotless = '_'.join([x.capitalize() for x in caps.split('.')])
+        return dotless
+
+    def _build_import_dicts(self, raw_conf, klass):
```
Comment (pjbreaux) [inline utils/RESTWalker.py]: This is only used for OrganizingCollections. Perhaps rename?

File: utils/RESTWalker.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+        imports = []
+        selfLinkstart = raw_conf[u"selfLink"].partition("?")[0]
```
Comment (pjbreaux) [inline utils/RESTWalker.py]: Does this fail if no '?' in link?

File: utils/RESTWalker.py [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+        items = raw_conf[u"items"]
+        for item in items:
+            if item[u"reference"][u"link"].startswith(selfLinkstart):
+                tempuri = item[u"reference"][u"link"]
+                pre_qm = tempuri.partition("?")[0]
+                post_sl = pre_qm[len(selfLinkstart)+1:]
+                caps = self._klassify(post_sl)
+                imports.append({'OC': '.'+klass.lower(),
+                                'module': '.'+caps.lower(),
+                                'klass': caps})
```
Comment (pjbreaux) [inline utils/RESTWalker.py]: Should we log or print out when item[reference][link] does not match selfLinkstart?

File: utils/RESTWalker.py [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+        return dotless
+
+    def _build_import_dicts(self, raw_conf, klass):
+        imports = []
+        selfLinkstart = raw_conf[u"selfLink"].partition("?")[0]
+        items = raw_conf[u"items"]
+        for item in items:
+            if item[u"reference"][u"link"].startswith(selfLinkstart):
+                tempuri = item[u"reference"][u"link"]
+                pre_qm = tempuri.partition("?")[0]
```
Comment (pjbreaux) [inline utils/RESTWalker.py]: pre_questionmark and post_selflink

File: utils/RESTWalker.py [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+                post_sl = pre_qm[len(selfLinkstart)+1:]
+                caps = self._klassify(post_sl)
```
Comment (pjbreaux) [inline utils/RESTWalker.py]: Call it 'classified'?

File: utils/RESTWalker.py [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+
+    def process_config(self, config_name):
+        raw_conf = self.raw_configs[config_name]
+        if 'kind' in raw_conf:
+            python_as_string = self._process_config_with_kind(raw_conf)
+        else:
+            raise Kindless(raw_conf)
+        return python_as_string
+
+    def _klassify(self, raw_klass):
```
Comment (pjbreaux) [inline utils/RESTWalker.py]: Rename to something like 'handle_dashes_dots_and_case'.

File: utils/RESTWalker.py [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+        template = self.templates['Resource']
+        python_as_string = template.render(container=container,
+                                           klass=raw_klass,
+                                           kind=kind)
+        return python_as_string
+
+    def _format_collection(self, kind, raw_conf):
+        raw_string = kind.split(':')[-1][:-len('collectionstate')]
+        raw_klass = self._klassify(raw_string)
+        if raw_klass.endswith('s'):
```
Comment (pjbreaux) [inline utils/RESTWalker.py]: Make its own function for readability and testability.

File: utils/RESTWalker.py [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+            coll_klass = raw_klass + '_s'
+        else:
+            coll_klass = raw_klass + 's'
+        member_klass = raw_klass
+        collection_template = self.templates['Collection']
+        memberkind = kind.replace('collection', '')
+        config_dict = {'coll_klass': coll_klass,
+                       'member_klass': member_klass,
+                       'collection_container': kind.split(':')[1],
```
Comment (pjbreaux) [inline utils/RESTWalker.py]: What if there are subcollections?

File: utils/RESTWalker.py [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+                       'collection_attr_reg_dict': {memberkind: member_klass}}
```
Comment (pjbreaux) [inline utils/RESTWalker.py]: For member_kind, what if you're a subcollection?

File: utils/RESTWalker.py [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+
+    def _format_stats(self, kind, raw_conf):
+        klass = kind.split(':')[-1][:-len('stats')]
+        print(klass)
+        template = self.templates['Stats']
+        print(template)
+        # XXX
+        # return python_as_string
+
+    def _process_config_with_kind(self, raw_conf):
```
Comment (pjbreaux) [inline utils/RESTWalker.py]: Identify why order matters in the docstring for your org_match etc...

File: utils/RESTWalker.py [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+        imps_as_list.sort()
+        imports = '\n'.join(imps_as_list)
+        python_as_string = self.license_template.render()+imports+OrgCollstr
+        self.formatted_configs = {klass: {'Python_str': python_as_string,
+                                          'config_dict': config_dict,
+                                          'import_dicts': import_dicts}}
+        return python_as_string
+
+    def _format_resource(self, kind, raw_conf):
+        raw_string = kind.split(':')[-1][:-len('state')]
```
Comment (pjbreaux) [inline utils/RESTWalker.py]: This resource name is often the second element from the end? Is it always the case? And do you want to know if that's not the case?

File: utils/RESTWalker.py [hunk 15]
```diff
@@ -0,0 +1,194 @@
...
+        raw_klass = self._klassify(raw_string)
+        if raw_klass.endswith('s'):
+            container = raw_klass+'_s'
+        else:
+            container = raw_klass+'s'
+        template = self.templates['Resource']
+        python_as_string = template.render(container=container,
```
Comment (pjbreaux) [inline utils/RESTWalker.py]: attr_reg_dict should be empty dict in output string.

File: utils/RESTWalker.py [hunk 16]
```diff
@@ -0,0 +1,194 @@
...
+    return obj
+
+
+def dump_json(obj, pathname):
+    fp = io.open(pathname, 'wb')
+    json.dump(obj, fp, indent=1)
+    fp.close()
+
+
+class TemplateEngine(object):
```
Comment (pjbreaux) [inline utils/RESTWalker.py]: This is puntable, but I'd like to see docstrings for the reasons why this thing exists. And possible info on the edge cases.

File: utils/RESTWalker.py [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+        os.path.walk(dirname, consumer, self)
+
+    def list_templates(self):
+        pp(self.templates.keys())
+
+    def list_raw_configs(self):
+        pp(self.raw_configs.keys())
+
+    def process_config(self, config_name):
+        raw_conf = self.raw_configs[config_name]
```
Comment (pjbreaux) [inline utils/RESTWalker.py]: If config_name not found (KeyError), tell them that it's a filename, and where it has to be located. All that jazz.

File: utils/RESTWalker.py (hunk 5 above)
Comment (zancas) [inline utils/RESTWalker.py]: I propose we punt this:  https://github.com/F5Networks/f5-common-python/issues/223

**Pull request merged**